db/pebble/db.go

```diff
@@ -87,6 +164,192 @@ func (d *DB) WithListener(listener db.EventListener) db.DB {
+	if d.activeComp == 0 {
+		d.compStartTime = time.Now()
+	}
+	l0 := info.Input[0]
+	if l0.Level == 0 {
+		d.level0Comp.Add(1)
+	} else {
+		d.nonLevel0Comp.Add(1)
+	}
+	d.activeComp++
+}
+
+func (d *DB) onCompactionEnd(info pebble.CompactionInfo) { //nolint:gocritic // Used by pebble's event listener
+	if d.activeComp == 1 {
+		d.compTime.Add(int64(time.Since(d.compStartTime)))
+	} else if d.activeComp == 0 {
+		panic("should not happen")
+	}
+	d.activeComp--
+}
+
+func (d *DB) onWriteStallBegin(b pebble.WriteStallBeginInfo) {
+	d.writeDelayStartTime = time.Now()
+}
+
+func (d *DB) onWriteStallEnd() {
+	d.writeDelayTime.Add(int64(time.Since(d.writeDelayStartTime)))
+}
+
+// collectMetrics periodically retrieves internal pebble counters and reports them to
+// the metrics subsystem.
+func (d *DB) StartMetricsCollection(ctx context.Context, refresh time.Duration) {
+	timer := time.NewTimer(refresh)
+	defer timer.Stop()
+
+	// Create storage and warning log tracer for write delay.
+	var (
+		compTimes        [2]int64
+		writeDelayTimes  [2]int64
+		writeDelayCounts [2]int64
+		compWrites       [2]int64
+		compReads        [2]int64
+
+		nWrites [2]int64
+	)
+
+	// Iterate ad infinitum and collect the stats
+	for i := 1; ; i++ {
+		var (
+			compWrite int64
+			compRead  int64
+			nWrite    int64
+
+			metrics            = d.pebble.Metrics()
+			compTime           = d.compTime.Load()
+			writeDelayCount    = d.writeDelayCount.Load()
+			writeDelayTime     = d.writeDelayTime.Load()
+			nonLevel0CompCount = int64(d.nonLevel0Comp.Load())
+			level0CompCount    = int64(d.level0Comp.Load())
+		)
+		writeDelayTimes[i%2] = writeDelayTime
+		writeDelayCounts[i%2] = writeDelayCount
+		compTimes[i%2] = compTime
+
+		for j := range metrics.Levels {
+			levelMetrics := metrics.Levels[j]
+			nWrite += int64(levelMetrics.BytesCompacted)
+			nWrite += int64(levelMetrics.BytesFlushed)
+			compWrite += int64(levelMetrics.BytesCompacted)
+			compRead += int64(levelMetrics.BytesRead)
+		}
+
+		nWrite += int64(metrics.WAL.BytesWritten)
+
+		compWrites[i%2] = compWrite
+		compReads[i%2] = compRead
+		nWrites[i%2] = nWrite
+
+		d.writeDelayCountMeter.Add(float64(writeDelayCounts[i%2] - writeDelayCounts[(i-1)%2]))
+		d.writeDelayTimeMeter.Add(float64(writeDelayTimes[i%2] - writeDelayTimes[(i-1)%2]))
+		d.compTimeMeter.Add(float64(compTimes[i%2] - compTimes[(i-1)%2]))
+		d.compReadMeter.Add(float64(compReads[i%2] - compReads[(i-1)%2]))
+		d.compWriteMeter.Add(float64(compWrites[i%2] - compWrites[(i-1)%2]))
+		d.diskSizeGauge.Set(float64(metrics.DiskSpaceUsage()))
+		d.diskWriteMeter.Add(float64(nWrites[i%2] - nWrites[(i-1)%2]))
+
+		// See https://github.com/cockroachdb/pebble/pull/1628#pullrequestreview-1026664054
+		manuallyAllocated := metrics.BlockCache.Size + int64(metrics.MemTable.Size) + int64(metrics.MemTable.ZombieSize)
+		d.manualMemAllocGauge.Set(float64(manuallyAllocated))
+		d.memCompGauge.Set(float64(metrics.Flush.Count))
```

Comment:
Is the fluch count the memory allocated for compaction??

Reply:
Made a mistake on the description of `memCompGauge`, actually `memCompGauge` tracks the number of memory allocations